blog/content/post/2022-12-22-fobnail_application.md

```diff
@@ -112,12 +118,242 @@ once. If any of the commands fails, further commands are not executed.
+Connect the Token and check if it gets its IP correctly, see [known issues](#known-issues)
+if not.
+
+With Token plugged in and with correct IP we can provision the platform. Fobnail
+Token Services are available even for Attester with provisioning functions, so
+encryption key can be written in the same invocation as platform provisioning.
+Following script automates the process of creating disk image, encryption key,
+platform provisioning and writing the encryption key to the Token. Change paths
+and names to your expectations, save it and run with `sudo`:
+
+```bash
+#!/bin/bash
+
+# Some configuration variables - change as needed
+DISK_IMG=/home/user/.secure_disk.img
+DISK_SIZE_MB=128
+MNT_DIR=/home/user/secure_storage
+ATTESTER=/usr/bin/fobnail-attester-with-provisioning
+MAPPER_DEV=c1
+TMP_KEY=/tmp/keyfile.bin
+FOBNAIL_KEY_FNAME=luks_key
+
+# Get user name, regardless of whether script is started with sudo or not
+DISK_USER=${SUDO_USER:-$USER}
+
+# Create disk image and mount is as loop device
+dd if=/dev/zero of=$DISK_IMG bs=1M count=$DISK_SIZE_MB
+LOOP_DEV=`losetup -f --show $DISK_IMG`
+
+# Create encryption key and LUKS partition
+dd bs=512 count=4 if=/dev/urandom of=$TMP_KEY
+cryptsetup luksFormat --type luks2 $LOOP_DEV $TMP_KEY
+
+# Create filesystem on the partition
+cryptsetup luksOpen -d $TMP_KEY $LOOP_DEV $MAPPER_DEV
+mke2fs -j /dev/mapper/$MAPPER_DEV
+
+# Mount partition and change the owner of root directory
+mkdir -p "$MNT_DIR"
+mount /dev/mapper/${MAPPER_DEV} "$MNT_DIR"
+chown -R $DISK_USER:$DISK_USER "$MNT_DIR"
+# Optionally fill the partition with secret stuff
+# cp top_secret_file.odt "${MNT_DIR}"
+umount "$MNT_DIR"
+
+# Close LUKS partition and underlying loop device
+cryptsetup close $MAPPER_DEV
+losetup -d $LOOP_DEV
+
+# Write key to Fobnail Token and securely erase it from host
+$ATTESTER --write-file $TMP_KEY:$FOBNAIL_KEY_FNAME && \
+    dd if=/dev/urandom of=$TMP_KEY bs=$(stat -c %s $TMP_KEY) count=1 && \
+    rm $TMP_KEY
+```
+
+Platform is now provisioned and encryption key can be read by calling
+`fobnail-attester --read-file luks_key:keyfile.bin`. For better user experience
+we can automate this as well with relatively simple scripts:
+
+- `mount.sh`:
+
+  ```bash
+  #!/bin/bash
+
+  # Always use full paths
+  DISK_IMG=/home/user/.secure_disk.img
+  MNT_DIR=/home/user/secure_storage
+  ATTESTER=/usr/bin/fobnail-attester
+  LOOP_DEV_FILE=/var/run/fobnail.loopdev
+  MAPPER_DEV=c1
+  FOBNAIL_KEY_FNAME=luks_key
+
+  # Create loop device and save its path for umount.sh
+  LOOP_DEV=`losetup -f --show $DISK_IMG`
+  echo -n "$LOOP_DEV" > "$LOOP_DEV_FILE"
+
+  # Give time for networkd service to set up Fobnail Token IP
+  sleep 5
+
+  # Get the key and open LUKS partition
+  timeout 30 $ATTESTER --read-file ${FOBNAIL_KEY_FNAME}:- | \
+      cryptsetup luksOpen -d - $LOOP_DEV $MAPPER_DEV
+
+  # Mount filesystem
+  mount /dev/mapper/${MAPPER_DEV} "$MNT_DIR"
+  ```
```

Comment:
@krystian-hebel please create repository with scripts in Fobnail organization on Github and use the scripts here instead of asking user for copy and pasting.

Reply:
MIT license is ok.

Reply:
Half of this is to be configured by users. I'm afraid that if these are saved as scripts most people won't even bother opening, much less changing the file contents.

For the future, when it will be more production-ready, I think we should change these scripts to proper systemd services, with separate configuration files, but that would be task for someone from EST probably.

Reply:
> Half of this is to be configured by users. I'm afraid that if these are saved as scripts most people won't even bother opening, much less changing the file contents.

Very complex configuration expected from users means lower adoption. Users need as seamless solution as possible.

> For the future, when it will be more production-ready, I think we should change these scripts to proper systemd services, with separate configuration files, but that would be task for someone from EST probably.

I don't know how you define "production-ready". If systemd is correctTM solution for the problem, which we not go with it now, is it really that hard that we need dedicated resources for implementing that?

IMHO if we will not care about quality of solution now, then we will most likely never improve it.

@macpijan I don't want to go into endless struggle about quality of this solution, if you agree with @krystian-hebel that copy&pasting scripts for users is good enough, then please resolve this conversation as "not an issue".

Reply:
Fobnail is a tool, these scripts are just examples of how it can be used. By analogy, when you install Thunderbird, you don't install scripts that automatically write emails.

Reply:
@krystian-hebel I'm not sure if I get analogy.  Research say that easiness of trying something in the project is key factor for community building. Thunderbird is email application there I don't know any user expecting from such application to automatically write emails. Please note Thunderbird have very easy to follow and automated on-boarding process for adding new account and starting to use it as email client to write, send and receive emails. We should try to provide such seamless experience for Fobnail.

GitHub knows quite a lot about building open source communities and they [recommend](https://opensource.guide/building-community/#setting-your-project-up-for-success) "Make it easy for someone to use your project.",  approaching user with above way IMO does not make it easy, and I'm ok if you have different opinion. Decision about leaving it as is or changing I leave to @macpijan .